### PR TITLE
KAFKA-2988: Change default configuration of the log cleaner

### PR DESCRIPTION
**Pull request by granthenke**

### CODE REVIEW

**Comment by ijuma** [comment]
We probably should mention this in the upgrade notes.

**Comment by hachikuji** [comment]
@granthenke Should we also change this in config/server.properties?

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 1]
```diff
@@ -80,13 +80,13 @@ object Defaults {
   val LogCleanupPolicy = Delete
   val LogCleanerThreads = 1
   val LogCleanerIoMaxBytesPerSecond = Double.MaxValue
-  val LogCleanerDedupeBufferSize = 500 * 1024 * 1024L
+  val LogCleanerDedupeBufferSize = 128 * 1024 * 1024L
```
Comment (junrao) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:82]: It's probably ok to lower this. Could you include this change in the update doc as Ismael mentioned?

File: core/src/main/scala/kafka/server/KafkaConfig.scala [hunk 2]
```diff
@@ -80,13 +80,13 @@ object Defaults {
...
   val LogCleanerIoBufferSize = 512 * 1024
   val LogCleanerDedupeBufferLoadFactor = 0.9d
   val LogCleanerBackoffMs = 15 * 1000
   val LogCleanerMinCleanRatio = 0.5d
-  val LogCleanerEnable = false
-  val LogCleanerDeleteRetentionMs = 24 * 60 * 60 * 1000L
+  val LogCleanerEnable = true
+  val LogCleanerDeleteRetentionMs = 7 * 24 * 60 * 60 * 1000L
```
Comment (junrao) [inline core/src/main/scala/kafka/server/KafkaConfig.scala]: Is there a particular reason to change this? This is really to just control how long we keep the delete tombstone. Basically, if a consumer has seen a deleted key, it's expected to consume the delete marker within this time. 1 day seems enough.

File: core/src/main/scala/kafka/server/KafkaConfig.scala (hunk 1 above)
Comment (granthenke) [inline core/src/main/scala/kafka/server/KafkaConfig.scala:82]: Are you saying this value is okay or should be lower?

**Comment by granthenke** [comment]
@ijuma I will add some updates to the upgrade notes

@junrao I changed log.cleaner.delete.retention.ms to 7 days to match the default log.retention time. I thought this would cause less "surprise" since the default SLA for "not missing any data" would then be the same between topics with cleanup.policy=delete and cleanup.policy=compact. Since removing the tombstones seamed like mostly a space saving optimization, I thought this change was fairly safe. I am happy to remove the change. Should I set it back to 1 day?

**Comment by granthenke** [comment]
@junrao Would you have time to review this?

File: docs/upgrade.html [hunk 3]
```diff
@@ -39,6 +39,7 @@
...
     <li> Configuration parameter replica.lag.max.messages was removed. Partition leaders will no longer consider the number of lagging messages when deciding which replicas are in sync. </li>
     <li> Configuration parameter replica.lag.time.max.ms now refers not just to the time passed since last fetch request from replica, but also to time since the replica last caught up. Replicas that are still fetching messages from leaders but did not catch up to the latest messages in replica.lag.time.max.ms will be considered out of sync. </li>
+    <li> Configuration parameter log.cleaner.enable is now true by default. This means topics with a cleanup.policy=compact will now be compacted by default, and 128 MiB of heap will be allocated to the cleaner process via log.cleaner.dedupe.buffer.size. You may want to review log.cleaner.dedupe.buffer.size and the other log.cleaner configuration values based on your usage of compacted topics. </li>
```
Comment (ijuma) [inline docs/upgrade.html]: Thanks for adding this. Is it worth updating `LogCleanerEnableDoc` to explain the implications of disabling that given that the new consumer relies on this for the offsets topic?

Comment (granthenke) [inline docs/upgrade.html]: @ijuma Yeah, I think thats a worthwhile addition. I will add something brief.

**Comment by granthenke** [comment]
@ijuma Updated the doc line

**Comment by junrao** [comment]
@granthenke : Sorry for the delay. Just a couple of minor comments.
1. log.cleaner.delete.retention.ms doesn't have to match log.retention. If a consumer has never seen any existing message for a key, it's ok for the consumer not to see the tombstone. It's only when a consumer has seen an existing message and the key is deleted, we expect the consumer to read the tombstone within some window. The current default of 1 day seems long enough. So, I recommend that we keep the current default.
2. Is 128 MiB a typo? Should it be 128MB?

**Comment by granthenke** [comment]
@junrao I will set log.cleaner.delete.retention.ms back to a day. 

The MiB wasn't a typo but a habit of mine. I understand its not necessarily used everywhere but its technically accurate: 

https://en.wikipedia.org/wiki/Mebibyte

> The unit was established by the International Electrotechnical Commission (IEC) in 1998. It was designed to replace the megabyte, which is still being used in many contexts to represent 2<sup>20</sup> bytes, which is incompatible with the definition of the prefix mega in the International System of Units (SI) as a multiplier of 10<sup>6</sup>.
> 
> The binary prefixes have been accepted by all major standards organizations and are part of the International System of Quantities. Many Linux distributions use the unit, but it is not yet widely published within the industry or media.

Happy to change it if you prefer to stick with MB.

**Comment by ijuma** [comment]
@granthenke I think we should be consistent with the rest of the code when it comes to MB versus MiB and it seems we use the former pretty much everywhere.

**Comment by granthenke** [comment]
@junrao @ijuma Rebased and updated based on comments

**Comment by junrao** [comment]
Thanks for the patch. LGTM

@ijuma : Do you know why the jenkins job didn't run the unit tests on this PR?

**Comment by ijuma** [comment]
@junrao Good question, I have no idea why it didn't run. Looks like a bug somewhere in the pipeline (doesn't seem to happen often).

**Comment by FreeSlaver** [comment]
Kafka,with a lot of bugs.Hard to use,your guys just show off  your High Technology,But not how to solve problems in reality.

**Pull request closed without merging**